cmd/util/ledger/migrations/cadence_values_migration_test.go

```diff
@@ -0,0 +1,138 @@
+package migrations
+
+import (
+	"testing"
+
+	"github.com/rs/zerolog"
+
+	_ "github.com/glebarez/go-sqlite"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+
+	"github.com/onflow/cadence/runtime/common"
+	"github.com/onflow/cadence/runtime/interpreter"
+	"github.com/onflow/cadence/runtime/sema"
+
+	"github.com/onflow/flow-go/cmd/util/ledger/reporters"
+	"github.com/onflow/flow-go/cmd/util/ledger/util"
+)
+
+const snapshotPath string = "test-data/cadence_values_migration/snapshot"
+
+func TestCadenceValuesMigration(t *testing.T) {
+
+	address, err := common.HexToAddress("01cf0e2f2f715450")
+	require.NoError(t, err)
+
+	// Get the old payloads
+
+	payloads, err := util.PayloadsFromEmulatorSnapshot(snapshotPath)
+	require.NoError(t, err)
+
+	// Migrate
+
+	rwf := &testReportWriterFactory{}
+	valueMigration := NewCadenceValueMigrator(rwf)
+
+	err = valueMigration.InitMigration(zerolog.Nop(), nil, 0)
+	require.NoError(t, err)
+
+	newPayloads, err := valueMigration.MigrateAccount(nil, address, payloads)
+	require.NoError(t, err)
+
+	// Assert the migrated payloads
+
+	mr, err := newMigratorRuntime(address, newPayloads)
+	require.NoError(t, err)
+
+	storageMap := mr.Storage.GetStorageMap(address, common.PathDomainStorage.Identifier(), false)
+	require.NotNil(t, storageMap)
+	require.Equal(t, 4, int(storageMap.Count()))
+
+	iterator := storageMap.Iterator(mr.Interpreter)
+
+	fullyEntitledAccountReferenceType := interpreter.ConvertSemaToStaticType(nil, sema.FullyEntitledAccountReferenceType)
+
+	var values []interpreter.Value
+	for key, value := iterator.Next(); key != nil; key, value = iterator.Next() {
+		// skip composite values. e.g: FlowToken etc.
+		if _, isComposite := value.(*interpreter.CompositeValue); isComposite {
+			continue
+		}
+		values = append(values, value)
+	}
+
+	// Order is non-deterministic, so use 'ElementsMatch'.
+	assert.ElementsMatch(
+		t,
+		values,
+		[]interpreter.Value{
+			// Both string values should be in the normalized form.
+			interpreter.NewUnmeteredStringValue("Caf\u00E9"),
+			interpreter.NewUnmeteredStringValue("Caf\u00E9"),
+
+			interpreter.NewUnmeteredTypeValue(fullyEntitledAccountReferenceType),
```

Comment:
Where are these values coming from?

Reply:
These values were constructed and stored in the emulator state (pre-1.0), for testing.
e.g. using the transaction:
```cadence
transaction {

  prepare(acct: AuthAccount) {
    acct.save("Cafe\u{0301}", to: /storage/string_value_1)
    acct.save("Caf\u{00E9}", to: /storage/string_value_2)
    acct.save(Type<AuthAccount>(), to: /storage/type_value_1)
  }

  execute {}
}
```

Reply:
Maybe commit the file, so that the test case can be changed

Reply:
It is there in [migrations/test-data/cadence_values_migration/snapshot](https://github.com/onflow/flow-go/blob/9eb4edb5fed4bccf5e8d763888dca54137ba9a38/cmd/util/ledger/migrations/test-data/cadence_values_migration/snapshot)

Reply:
Yeah, the transaction Cadence code that was used to generate the snapshot